ambry-utils/src/main/java/com.github.ambry.utils/ByteBufChannel.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.utils;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.WritableByteChannel;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * A {@link WritableByteChannel} that stores the bytes written into it in a {@link ByteBuf}.
+ */
+public class ByteBufChannel implements WritableByteChannel {
+  private final AtomicBoolean channelOpen = new AtomicBoolean(true);
+  private CompositeByteBuf compositeBuffer = ByteBufAllocator.DEFAULT.compositeBuffer();
+
+  /**
+   * Gets the {@link CompositeByteBuf} that is being used to receive writes.
+   * @return the {@link ByteBuf} that is receives writes to this channel.
+   */
+  public ByteBuf getBuf() {
+    return compositeBuffer;
+  }
+
+  /**
+   * The default constructor.
+   */
+  public ByteBufChannel() {
+
+  }
+
+  /**
+   * Copies bytes from {@code src} into the composite buf.
+   * @param src the source {@link ByteBuf} to copy bytes from.
+   * @return the number of bytes copied.
+   * @throws ClosedChannelException if the channel is closed when this function was called.
+   */
+  @Override
+  public int write(ByteBuffer src) throws ClosedChannelException {
+    int size = src.remaining();
+    compositeBuffer.addComponent(true, Unpooled.wrappedBuffer(src));
```

Comment:
Generally for implementations of `WritableByteChannel`, I think you are supposed to copy the bytes instead of just wrapping them since `ByteBuffer`s have no ref count system and may be reused by the writer afterwards. (This is what many of the code snippets for copying from ReadableByteChannel to WritableByteChannel do).

I think for the router, it may be okay to keep this optimization present but I did find some cases that I was a little unsure about. The suspicious case I found was PutOperation where the buffer is cleared in `prepareForBuilding`. I think this method is only called after a chunk is cleared. This happens when a chunk operation is completed, so for requests that we care about, it hopefully won't cause issues, but a chunk operation being completed does not necessarily mean all requests have been successfully sent out (some could be stuck in the queue).

Reply:
I noticed some PUT issue in EI test. Not sure yet.

To make everything safe, the latest change use copy in ByteBufChannel. Next step is to use Use  RetainingAsyncWritableChannel after PUT/GET request's `writeTo(AsyncWritableChannel channel, Callback<Long> callback)` is fully implemented.